File: src/factories/MetadataTypeTagSchemaFactory.ts
import { Metadata } from "../schemas/metadata/Metadata";
import { MetadataObjectType } from "../schemas/metadata/MetadataObjectType";

export namespace MetadataTypeTagSchemaFactory {
  export const object = (props: {
    report: (msg: string) => false;
    object: MetadataObjectType;
  }): object | undefined => {
    if (props.object.recursive) {
      props.report(`${props.object.name} has recursive type`);
      return undefined;
    }
    const output: any = {};
    for (const p of props.object.properties) {
      const key: string | null = p.key.getSoleLiteral()!;
      if (key === null) {
        props.report(
          `${props.object.name} has non-literal key type: ${p.key.getName()}`,
        );
        continue;
      }
      output[key] = iterate({
        report: props.report,
        object: props.object,
        key,
        metadata: p.value,
      });
    }
    return output;
  };

  const iterate = (props: {
    report: (message: string) => false;
    object: MetadataObjectType;
    key: string;
    metadata: Metadata;
  }): any => {
    if (
      props.metadata.any ||
      props.metadata.atomics.length ||
      props.metadata.arrays.length ||
      props.metadata.natives.length ||
      props.metadata.functions.length
    )
      props.report(`${props.object.name}.${props.key} has non-literal type`);
    else if (props.metadata.size() > 1)
      props.report(`${props.object.name}.${props.key} has union type`);
    else if (props.metadata.size() === 0)
      if (props.metadata.nullable) return null;
      else if (props.metadata.isRequired() === true)
        props.report(`${props.object.name}.${props.key} has non-literal type`);
      else return undefined;
    else if (props.metadata.constants.length)
      return props.metadata.constants[0]!.values[0]!.value;
    else if (props.metadata.tuples.length) {
      const tuple = props.metadata.tuples[0]!;
      if (tuple.type.isRest())
        props.report(`${props.object.name}.${props.key} has rest tuple type`);
      else if (tuple.type.recursive)
        props.report(
          `${props.object.name}.${props.key} has recursive tuple type`,
        );
      else if (tuple.type.elements.some((e) => e.required === false))
        props.report(
          `${props.object.name}.${props.key} has optional tuple type`,
        );
      return tuple.type.elements.map((elem) =>
        iterate({
          report: props.report,
          object: props.object,
          key: props.key,
          metadata: elem,
        }),
      );
    } else if (props.metadata.objects.length)
      return object({
        report: props.report,
        object: props.metadata.objects[0]!.type,
      });
    else props.report(`${props.object.name}.${props.key} has non-literal type`);
  };
}

File: packages/typia/native/core/factories/MetadataTypeTagSchemaFactory.go
package factories

import nativemetadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"

type metadataTypeTagSchemaFactoryNamespace struct{}

var MetadataTypeTagSchemaFactory = metadataTypeTagSchemaFactoryNamespace{}

func (metadataTypeTagSchemaFactoryNamespace) Object(props struct {
  Report func(msg string) bool
  Object *nativemetadata.MetadataObjectType
}) map[string]any {
  if props.Object.Recursive {
    props.Report(props.Object.Name + " has recursive type")
    return nil
  }
  output := map[string]any{}
  for _, property := range props.Object.Properties {
    key := property.Key.GetSoleLiteral()
    if key == nil {
      props.Report(props.Object.Name + " has non-literal key type: " + property.Key.GetName())
      continue
    }
    value, ok := metadataTypeTagSchemaFactory_iterate(struct {
      report   func(message string) bool
      object   *nativemetadata.MetadataObjectType
      key      string
      metadata *nativemetadata.MetadataSchema
    }{
      report:   props.Report,
      object:   props.Object,
      key:      *key,
      metadata: property.Value,
    })
    if ok {
      output[*key] = value
    }
  }
  return output
}

func metadataTypeTagSchemaFactory_iterate(props struct {
  report   func(message string) bool
  object   *nativemetadata.MetadataObjectType
  key      string
  metadata *nativemetadata.MetadataSchema
}) (any, bool) {
  if props.metadata.Any ||
    len(props.metadata.Atomics) != 0 ||
    len(props.metadata.Arrays) != 0 ||
    len(props.metadata.Natives) != 0 ||
    len(props.metadata.Functions) != 0 {
    props.report(props.object.Name + "." + props.key + " has non-literal type")
    return nil, true
  }
  if props.metadata.Size() > 1 {
    props.report(props.object.Name + "." + props.key + " has union type")
    return nil, true
  }
  if props.metadata.Size() == 0 {
    if props.metadata.Nullable {
      return nil, true
    }
    if props.metadata.IsRequired() {
      props.report(props.object.Name + "." + props.key + " has non-literal type")
      return nil, true
    }
    return nil, false
  }
  if len(props.metadata.Constants) != 0 {
    return props.metadata.Constants[0].Values[0].Value, true
  }
  if len(props.metadata.Tuples) != 0 {
    tuple := props.metadata.Tuples[0]
    if tuple.Type.IsRest() {
      props.report(props.object.Name + "." + props.key + " has rest tuple type")
    } else if tuple.Type.Recursive {
      props.report(props.object.Name + "." + props.key + " has recursive tuple type")
    } else {
      for _, elem := range tuple.Type.Elements {
        if elem.Required == false {
          props.report(props.object.Name + "." + props.key + " has optional tuple type")
          break
        }
      }
    }
    array := []any{}
    for _, elem := range tuple.Type.Elements {
      value, ok := metadataTypeTagSchemaFactory_iterate(struct {
        report   func(message string) bool
        object   *nativemetadata.MetadataObjectType
        key      string
        metadata *nativemetadata.MetadataSchema
      }{
        report:   props.report,
        object:   props.object,
        key:      props.key,
        metadata: elem,
      })
      if ok {
        array = append(array, value)
      }
    }
    return array, true
  }
  if len(props.metadata.Objects) != 0 {
    return MetadataTypeTagSchemaFactory.Object(struct {
      Report func(msg string) bool
      Object *nativemetadata.MetadataObjectType
    }{
      Report: props.report,
      Object: props.metadata.Objects[0].Type,
    }), true
  }
  props.report(props.object.Name + "." + props.key + " has non-literal type")
  return nil, true
}
